### DISCUSSION TITLE
Holding on to a websocket

### DISCUSSION THREAD

**Discussion by MatanRubin** [fastapi/fastapi#7918] [Questions] [answered]
### First check

* [V] I used the GitHub search to find a similar issue and didn't find it.
* [V] I searched the FastAPI documentation, with the integrated search.
* [V] I already searched in Google "How to X in FastAPI" and didn't find any information.

### Description
I have an MQTT client receiving messages from an external server. 
I want to forward each received message to my client using a websocket.
To do that, I want my client to first connect to a FastAPI websocket and every time a message comes in from the MQTT server, I want to `sent_text` on the websocket.

I tried something like the following, but couldn't get it to work. What am I missing?

```python
client_websocket = None 

@app.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket):
    await websocket.accept()
    global client_websocket
    client_websocket = websocket

async def on_mqtt_message(msg):
    await client_websocket.sent_text(msg)
```

This raises the following error: `RuntimeError: Unexpected ASGI message 'websocket.send', after sending 'websocket.close'.`

From the error I understand that the websocket is closed once `websocket_endpoint` exits. Is there a way to keep the websocket open, and emit messages to it from other contexts?

The websocket examples in the doc all use `while True` to read data from the websocket and then respond, but in my case I have an extrenal system trigerring events which should create websocket messages.

I'd appreciate some help in understanding how to properly work with websockets in FastAPI (which is absolutely awesome, BTW).

Thanks!

**Comment by Dustyposa** [comment]
You expected multiple client use the one `socket`?

**Comment by michaelyaakoby** [comment, marked as answer]
You can probably use an async queue so your MQTT client will push messages to the queue and the WS server will get from the queue and send them to the WS client.
```python
from asyncio import Queue

queue: Queue = None

@app.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket):
    await websocket.accept()
    global queue
    queue = Queue()
    while True:
        msg = await queue.get()
        await websocket.send_text({"message": msg})


async def on_mqtt_message(msg):
    if queue:
      await queue.put(msg)
```
Note that the queue must be using the same event-loop as the WS, one way of doing it is to initialize the queue in the same function where you accept the WS.
Also, assuming you want the WS client to receive all the MQTT messages, you should use a queue per WS.

**Comment by MatanRubin** [comment]
Thanks @michaelyaakoby , that did the trick!

**Comment by tiangolo** [comment]
Thanks for the help here everyone! :cake: :bowing_man: 

And thanks @MatanRubin for reporting back and closing the issue :+1:
